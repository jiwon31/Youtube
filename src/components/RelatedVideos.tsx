import { useQuery } from "@tanstack/react-query";
import { useYoutubeApi } from "context/YoutubeApiContext";
import VideoCard from "./VideoCard";

type RelatedVideosProps = { id: string };

export default function RelatedVideos({ id }: RelatedVideosProps) {
  const { youtube } = useYoutubeApi();
  const {
    isLoading,
    error,
    data: videos,
  } = useQuery(["related", id], async () => youtube.relatedVideos(id), {
    staleTime: 1000 * 60 * 5,
  });

  return (
    <>
      {isLoading && <p>Loading...</p>}
      {error && <p>Something is wrong ðŸ˜£</p>}
      <ul>
        {videos &&
          videos.map((video) => (
            <VideoCard key={video.id} video={video} type="list" />
          ))}
      </ul>
    </>
  );
}
