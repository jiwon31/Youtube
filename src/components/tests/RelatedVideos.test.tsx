import {
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import RelatedVideos from "components/RelatedVideos";
import { withAllContext, withRouter } from "tests/utils";
import { Route } from "react-router-dom";
import { fakeVideos as videos } from "tests/videos";

describe("RelatedVideos", () => {
  const fakeYoutube = {
    relatedVideos: jest.fn(),
  };
  afterEach(() => fakeYoutube.relatedVideos.mockReset());

  it("renders", async () => {
    fakeYoutube.relatedVideos.mockImplementation(() => videos);
    const { asFragment } = renderRelatedVideos();

    // 로딩이 없어질 때까지 기다렸다가
    await waitForElementToBeRemoved(screen.queryByText("Loading..."));
    // await screen.findAllByRole("listitem")
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders related videos", async () => {
    fakeYoutube.relatedVideos.mockImplementation(() => videos);
    renderRelatedVideos();

    expect(fakeYoutube.relatedVideos).toHaveBeenCalledWith("1");
    await waitFor(() =>
      expect(screen.getAllByRole("listitem")).toHaveLength(videos.length)
    );
  });

  it("renders error", async () => {
    fakeYoutube.relatedVideos.mockImplementation(() => {
      throw new Error("error");
    });
    renderRelatedVideos();

    await screen.findByText("Something is wrong 😣");
  });

  it("renders loading", () => {
    fakeYoutube.relatedVideos.mockImplementation(() => videos);
    renderRelatedVideos();

    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  function renderRelatedVideos() {
    return render(
      withAllContext(
        withRouter(<Route path="/" element={<RelatedVideos id="1" />} />),
        fakeYoutube
      )
    );
  }
});
