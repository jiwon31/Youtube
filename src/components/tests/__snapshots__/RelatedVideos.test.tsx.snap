// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedVideos renders 1`] = `
<DocumentFragment>
  <ul>
    <li
      class="flex gap-2 m-2 ml-3 hover:scale-105 ease-in duration-200 cursor-pointer"
    >
      <img
        alt="title"
        class="w-40"
        src="http://image/"
      />
      <div>
        <p
          class="text-sm font-semibold my-2 line-clamp-2"
        >
          title
        </p>
        <p
          class="text-xs opacity-80"
        >
          channelTitle
        </p>
        <p
          class="text-xs opacity-80"
        >
          3 weeks ago
        </p>
      </div>
    </li>
    <li
      class="flex gap-2 m-2 ml-3 hover:scale-105 ease-in duration-200 cursor-pointer"
    >
      <img
        alt="title2"
        class="w-40"
        src="http://image2/"
      />
      <div>
        <p
          class="text-sm font-semibold my-2 line-clamp-2"
        >
          title2
        </p>
        <p
          class="text-xs opacity-80"
        >
          channelTitle2
        </p>
        <p
          class="text-xs opacity-80"
        >
          3 weeks ago
        </p>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
